<a href="https://colab.research.google.com/github/N1colasFD/Pruebas_Tecnicas_DS/blob/main/EricssonLatam_P_tecnica_Nicolas_fajardo01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
# Si algo no funciona, envienme un correo :)

# Ericsson Prueba Tecnica para Latino-America

1. Para esta prueba técnica tiene que calcular la rentabilidad de un capital de 20.000 y 40.000 Euros para un promedio de 3000 personas
que hayan invertido entre 3% y 8%  anual durante 3 años; **NOTA**: debe priorizar clientes de 4% anual que será nuestro enfoque comercial 'PPP'.

2. Aplicando la fórmula de 'interes_simple' y 'interes_compuesto' cree la condiciones para clasificación estándar Europea de clientes 'A','AA','AAA'

3. filtre la tabla de acuerdo a las clasificaciones correctas

4. Ahora deberá poder almacenar manualmente el interés compuesto y cada celda debe actualizarse correctamente.
EJ: imaginemos que queremos depositar 10.000 euros a un 3% anual de interés compuesto para cada cliente y queremos calcular el monto de intereses generado en dos años. **NOTA**: para este ejercicio debe establecer con métodos de usuario para posterior entrar almacenarlo en Bases de datos de Microsoft Azure y AWS

# A Tomar  en cuenta
1. El código debe ejecutarse con un solo `comando` 
2. Se valorará el uso de buenas prácticas en programación
3. Debe usar uno y solo un gráfico de su elección para mostrar los resultados 
4. considere buen uso de prácticas para empresa
5. *No* comente su código
6. generar marcas de agua

# Asegurase de:

En esta parte habia condicciones por la empresa que no viene al caso exponerlas aca, asi que las borre*



In [1]:
#Paquetes basicos
import pandas as pd
import seaborn as sns
import numpy as np
import random
import matplotlib.pyplot as plt
#paquetes avanzados
%matplotlib inline
plt.rcParams['figure.figsize'] = (10,4)
import statistics 

In [2]:
#Tiempos aleatorios 
from datetime import datetime, timedelta
#visualizaciones dinamicas
import plotly.express as px

In [3]:
#Autocompletar rápido 🚀
%config IPCompleter.greedy=True

#Display 6 registros 🚦
pd.options.display.min_rows = 4

#Formato de Notacion Cientifica 🛰
pd.options.display.float_format = '{:15.2f}'.format

-------

In [11]:
np.random.seed(123)

def clasificar_cliente(capital, edad, interes_simple):
    if edad > 60 and capital < 30000:
        return 'A'
    elif 25000 <= capital <= 30000 and 18 <= edad <= 70:
        return 'AA'
    elif capital > 35000 and interes_simple >= 0.06:
        return 'AAA'
    elif interes_simple == 0.04:
        return 'PPP'
    else:
        return 'SC'

n = 3000
capital = np.random.randint(20000, 40000, n)
interes_simple = np.random.uniform(0.03, 0.08, n)
edad = np.random.randint(18, 70, n)
saldo_3_anos_simple = capital * interes_simple * 3
saldo_3_anos_compuesto = capital * np.power(1 + interes_simple, 3)
rentabilidad = saldo_3_anos_compuesto - capital
tipo_cliente = [clasificar_cliente(c, e, i) for c, e, i in zip(capital, edad, interes_simple)]


df = pd.DataFrame({'capital': capital, 'interes_simple': interes_simple, 
                   'edad': edad, 'saldo_3_anos_simple': saldo_3_anos_simple, 
                   'saldo_3_anos_compuesto': saldo_3_anos_compuesto,
                   'rentabilidad':rentabilidad,
                   'tipo_cliente': tipo_cliente})

df[['capital', 'saldo_3_anos_simple', 'saldo_3_anos_compuesto', 'rentabilidad']] = \
    df[['capital', 'saldo_3_anos_simple', 'saldo_3_anos_compuesto', 'rentabilidad']].round(2)

fecha_inicial_min = pd.Timestamp(datetime(2022, 1, 1))
fecha_inicial_max = pd.Timestamp(datetime(2022, 12, 31))
diferencia = fecha_inicial_max - fecha_inicial_min
fechas_inicio = [fecha_inicial_min + pd.Timedelta(days=random.randint(0, diferencia.days)) for i in range(3000)]
df['fecha_inicio'] = fechas_inicio
df['fecha_fin'] = df['fecha_inicio'] + timedelta(days=3*365)

df.sample(n = 50)


capital  interes_simple  edad  saldo_3_anos_simple  \
691     22777            0.03    44              2379.61   
1106    36625            0.05    30              5652.00   
1753    34789            0.03    54              3269.92   
1207    27685            0.08    68              6519.75   
2671    29996            0.05    50              4673.09   
986     24788            0.07    38              5294.70   
1660    29998            0.06    18              4959.93   
1069    28327            0.04    39              3341.41   
1282    37337            0.04    36              4336.58   
873     39228            0.04    38              5156.41   
1035    39318            0.04    52              5040.38   
1455    25875            0.07    43              5687.85   
2140    22250            0.05    41              3215.28   
1240    21209            0.05    22              3449.85   
320     39679            0.07    20              8056.48   
1724    38307            0.05    54              5960.53   
2599    26833            0.07    45              5879.12   
2498    32980            0.04    41              3693.55   
1599    39754            0.05    65              6024.99   
2204    26970            0.05    27              3916.21   
2135    39768            0.06    26              7714.70   
2846    34948            0.03    50              3265.65   
2010    23560            0.05    21              3851.52   
414     22201            0.07    25              4819.39   
472     25000            0.08    19              5817.83   
2717    25411            0.06    64              4606.78   
2188    39253            0.06    67              6778.46   
2003    36026            0.05    24              4984.65   
171     27044            0.06    54              5055.89   
1995    23997            0.04    56              3185.53   
1483    26370            0.08    33              6152.59   
1587    35377            0.06    57              6700.32   
19      26484            0.03    21              2768.76   
1337    39495            0.03    51              3616.05   
2650    30353            0.04    26              3375.72   
1388    30426            0.08    69              6884.05   
325     35763            0.07    62              7566.94   
2597    29931            0.04    52              3588.86   
2903    38697            0.08    31              8936.58   
421     32871            0.04    22              3842.84   
2570    22958            0.04    33              2858.30   
954     22043            0.04    21              2334.63   
360     39162            0.08    31              9285.74   
289     26086            0.07    55              5830.73   
1643    25745            0.06    57              4469.51   
2430    36885            0.07    50              8269.59   
1534    23577            0.04    61              2793.00   
1096    29676            0.06    51              5571.53   
1495    23486            0.07    60              4852.67   
1184    29739            0.04    63              3271.99   

      saldo_3_anos_compuesto    rentabilidad tipo_cliente fecha_inicio  \
691                 25240.45         2463.45           SC   2022-02-07   
1106                42572.73         5947.73           SC   2022-01-06   
1753                38162.43         3373.43           SC   2022-02-07   
1207                34729.93         7044.93            A   2022-08-12   
2671                34915.97         4919.97           AA   2022-09-25   
986                 30468.63         5680.63           SC   2022-02-23   
1660                35236.32         5238.32           AA   2022-02-19   
1069                31801.52         3474.52           AA   2022-03-15   
1282                41843.64         4506.64           SC   2022-06-02   
873                 44613.64         5385.64           SC   2022-12-11   
1035                44576.83         5258.83           SC   2022-06-22   
1455                31989.80         6114.80           AA

In [12]:
df

capital  interes_simple  edad  saldo_3_anos_simple  \
0       39966            0.07    50              8826.69   
1       35725            0.04    47              4289.39   
...       ...             ...   ...                  ...   
2998    32067            0.04    46              3595.63   
2999    26134            0.06    69              4352.14   

      saldo_3_anos_compuesto    rentabilidad tipo_cliente fecha_inicio  \
0                   49458.44         9492.44          AAA   2022-10-17   
1                   40188.35         4463.35           SC   2022-01-02   
...                      ...             ...          ...          ...   
2998                35798.69         3731.69           SC   2022-01-31   
2999                30732.20         4598.20            A   2022-12-19   

      fecha_fin  
0    2025-10-16  
1    2025-01-01  
...         ...  
2998 2025-01-30  
2999 2025-12-18  

[3000 rows x 9 columns]

In [13]:
def agregar_registro(df, index):
    capital = int(input("Ingrese el capital: "))
    interes_simple = float(input("Ingrese el interés simple: "))
    edad = int(input("Ingrese la edad: "))
    saldo_3_anos_simple = capital * (1 + 3 * interes_simple)
    saldo_3_anos_compuesto = capital * (1 + interes_simple) ** 3
    rentabilidad = saldo_3_anos_compuesto - capital
    fecha_inicio = pd.Timestamp(datetime.now())
    fecha_fin = fecha_inicio + timedelta(days=3*365)
    tipo_cliente = clasificar_cliente(capital, edad, interes_simple)
    df.loc[index] = [capital, interes_simple, edad, saldo_3_anos_simple, saldo_3_anos_compuesto, fecha_inicio, fecha_fin, tipo_cliente]

df = pd.DataFrame(columns=['capital', 'interes_simple', 'edad', 'saldo_3_anos_simple', 'saldo_3_anos_compuesto','fecha_inicio', 'fecha_fin', 'tipo_cliente'])
def agregar_multiples_registros(df, num_registros):
    df['fecha_inicio'] = None
    for i in range(0, 1 + num_registros):
        agregar_registro(df, i)
agregar_multiples_registros(df, 2)


Ingrese el capital: 65265
Ingrese el interés simple: 0.04
Ingrese la edad: 26
Ingrese el capital: 100
Ingrese el interés simple: 0.04
Ingrese la edad: 10
Ingrese el capital: 20
Ingrese el interés simple: 0.08
Ingrese la edad: 30


In [14]:
df

capital  interes_simple edad  saldo_3_anos_simple  saldo_3_anos_compuesto  \
0   65265            0.04   26             73096.80                73414.25   
1     100            0.04   10               112.00                  112.49   
2      20            0.08   30                24.80                   25.19   

                fecha_inicio                  fecha_fin tipo_cliente  
0 2023-02-14 04:56:46.447470 2026-02-13 04:56:46.447470          PPP  
1 2023-02-14 04:56:57.564345 2026-02-13 04:56:57.564345          PPP  
2 2023-02-14 04:57:23.382576 2026-02-13 04:57:23.382576           SC

# Para quien quiera realizar las siguientes verificaciones

1. Concatene las Tablas; **Nota**: Esto le generará unos Nulos y otros problemas dependiendo de como utilice la concatenación
    Fíjese muy bien en que falta y lo **NO** falta, este es el paso final de la Prueba técnica para una empresa llamada Ericsson 
    dedica a resolver problemas de infraestructura tecnológica a muchas empresas.

2.  La oferta laboral era para un científico de datos para la resolución de problemas de este tipo
    por ende si ustedes es capaz de resolver el paso final de esta prueba puede agregarlo a su portafolio para y agregarlo a su HV

3. tenga muy en cuenta que las fechas le generan problemas de todo tipo, pero existe un parámetro universal para arreglarlo.

4. Una de mis principales fuentes de ingresos es la concatenación de tablas mal dimensionadas, aproximadamente 40 a la semana, 
    dichas tablas viene de empresas ubicadas en Europa, las recibo por medio de mis compañeros en el máster de Ml que estoy haciendo 
    En España existen al rededor de 17 mil ofertas laborales para DS y DA la mayoría pide portafolio

5. Como nota final recuerde que debe generar un gráfico para explicar la tabla de la mejor forma posible,
    con el tiempo la publicaré en redes, pero quisiera ver la habilidad de las personas...
    Si todo sale bien, empezaré a trabajar con una empresa la cual requiere un equipo de trabajo aca en Colombia
    por ende si usted desea puede enviarme el paso final de la prueba mi correo: nicolasfajardodaza@gmail.com con todo su análisis

6. Recuerde que si desea usar ChatGpt y similares no tiene ningun problema, pero quisiera que por favor me lo indicara. 

# Extra 
Si usted es Científico de datos, esta prueba técnica la recibí el mes pasado y aproximante la publicaré, si usted desea también envíemela a mi correo, como dije anteriormente, recibo bastante trabajo de este tipo: 

Un inversionista Americano residente en Colombia adquiere un título que tiene un valor de 3.500
US$, gana un interés del 4.5% y tiene un plazo de un año, pero no genera intereses los fines de semana y festivos, la tasa de cambio actual es de $ 1.500/US$ y se estima una devaluación en ese año del 19%. Calcular la rentabilidad
que obtendrá el inversionista en un DF con los 365 días. 
1. Use uno y solo un gráfico explicativo 
2. buenas prácticas de código
3. si usa ChatGpt indíquelo (esto no tiene nada de malo)